### DISCUSSION TITLE
Question on vector type for FFT

### DISCUSSION THREAD

**Discussion by arnaudDewetronfrance** [DEWETRON/OXYGEN-SDK#21] [Q&A] [answered]
Dear all , 

I have a question about vector data format. 

Is it possible to create into the SDK a vector channel that can be used in FFT virtual instrument ? 

Regards

**Comment by matthiasstraka** [comment, marked as answer]
Hello!

Yes, it is possible to create a vector channel from an SDK plugin that can be displayed in the FFT instrument (Spectrum analyzer). The only requirement is that the output channel is a SYNC, scalar channel which has a property `VectorDescription` of type `odk::Range`. The range property assigns a linear axis to the vector channel dimensions.

The following code will create a compatible channel:
```cpp
auto ch = addOutputChannel("fft_displayable_channel", getRootChannel());
ch->setSampleFormat(odk::ChannelDataformat::SampleOccurrence::SYNC, odk::ChannelDataformat::SampleFormat::DOUBLE, 10); // vector with 10 double elements
ch->addProperty("VectorDescription", odk::Property("", odk::Range(0, 9, "Hz"))); // Create labels for the bins (0 Hz, 1 Hz, ... 9 Hz)
ch->setDefaultName("VectorChannel")
   .setSimpleTimebase(10000)
   .setRange(odk::Range(0, 100, "V"))
   .setUnit("V");
```

When emitting samples to this channel, use a code similar to this:
```cpp
double values[10] = { 1,2,3,4,5,6,7,8,9,0 };
auto ch = getOutputChannelByKey("fft_displayable_channel");
odk::addSample(host, ch->getLocalId(), 0, values, sizeof(values));
```

Let us know if that solves your request.

**Comment by arnaudDewetronfrance** [comment]
Dear Matthias,

Thank for this response , I have tested and no problem.

I have just a question about frequency value setting , is there an other methods that use “range(0,9,Hz)” especially a way to define double value for it and not only UNIT representation ?

General topic , do you a document that explain all the property and methods that can be used into the SDK ?

I remember a case in the first version where I tried to create dynamic channel into my plugin ( user select for example into the root channel , the number of channel to create and these one are created according the user actions or deleted ) and
I faced some issues with the property of these news dynamics channels . When the dms file was for example reloaded or switch from replay to active session, the property of these channes was lost.

Do you know if there a way to manage it better in new SDK version ?

Regards

De : Matthias Straka ***@***.***>
Envoyé : mardi 6 décembre 2022 08:48
À : DEWETRON/OXYGEN-SDK ***@***.***>
Cc : Arnaud Genebrier ***@***.***>; Author ***@***.***>
Objet : Re: [DEWETRON/OXYGEN-SDK] Question on vector type for FFT (Discussion #21)

Vous ne recevez pas souvent de courriers de la part de ***@***.******@***.***>. Découvrez pourquoi cela est important<https://aka.ms/LearnAboutSenderIdentification>

This email originated outside of DEWETRON. DO NOT CLICK links or attachments unless you recognize the sender and know the content is safe.

Hello!

Yes, it is possible to create a vector channel from an SDK plugin that can be displayed in the FFT instrument (Spectrum analyzer). The only requirement is that the output channel is a SYNC, scalar channel which has a property VectorDescription of type odk::Range. The range property assigns a linear axis to the vector channel dimensions.

The following code will create a compatible channel:

auto ch = addOutputChannel("fft_displayable_channel", getRootChannel());

ch->setSampleFormat(odk::ChannelDataformat::SampleOccurrence::SYNC, odk::ChannelDataformat::SampleFormat::DOUBLE, 10); // vector with 10 double elements

ch->addProperty("VectorDescription", odk::Property("", odk::Range(0, 9, "Hz"))); // Create labels for the bins (0 Hz, 1 Hz, ... 9 Hz)

ch->setDefaultName("VectorChannel")

   .setSimpleTimebase(10000)

   .setRange(odk::Range(0, 100, "V"))

   .setUnit("V");

When emitting samples to this channel, use a code similar to this:

double values[10] = { 1,2,3,4,5,6,7,8,9,0 };

auto ch = getOutputChannelByKey("fft_displayable_channel");

odk::addSample(host, ch->getLocalId(), 0, values, sizeof(values));

Let us know if that solves your request.

**Comment by matthiasstraka** [comment]
> is there an other methods that use “range(0,9,Hz)” especially a way to define double value for it and not only UNIT representation ?

The `Range()` constructor takes two double values at the moment, the two integer values will get casted to double automatically but any other double value should be fine.

> General topic , do you a document that explain all the property and methods that can be used into the SDK ?

There is documentation available at https://dewetron.github.io/OXYGEN-SDK/
Browsing through the code examples is the best way to see what is possible at the moment and a good guideline for the right order of commands. It is best to look at code examples in the latest SDK version (e.g. on [Github](https://github.com/DEWETRON/OXYGEN-SDK/tree/master/examples)) there were many improvements and simplifications throughout recent releases.

If you have a specific problem, maybe you can write a simple example plugin that shows your problem and open an issue so we can take a look.
